beacon-chain/sync/validate_sync_committee_message.go

```
@@ -0,0 +1,310 @@
+
+		// Ignore the message if it is not possible to retrieve the signing root.
+		d, err := s.cfg.Chain.HeadSyncCommitteeDomain(ctx, m.Slot)
+		if err != nil {
+			traceutil.AnnotateError(span, err)
+			return pubsub.ValidationIgnore
+		}
+		rawBytes := p2ptypes.SSZBytes(m.BlockRoot)
+		sigRoot, err := helpers.ComputeSigningRoot(&rawBytes, d)
+		if err != nil {
+			traceutil.AnnotateError(span, err)
+			return pubsub.ValidationIgnore
+		}
+
+		// Reject for a validator index that is not found.
+		pubKey, err := s.cfg.Chain.HeadValidatorIndexToPublicKey(ctx, m.ValidatorIndex)
+		if err != nil {
+			traceutil.AnnotateError(span, err)
+			return pubsub.ValidationReject
+		}
+
+		// We reject a malformed signature from bytes according to the p2p specification.
+		blsSig, err := bls.SignatureFromBytes(m.Signature)
+		if err != nil {
+			traceutil.AnnotateError(span, err)
+			return pubsub.ValidationReject
+		}
+
+		// Ignore a malformed public key from bytes according to the p2p specification.
+		pKey, err := bls.PublicKeyFromBytes(pubKey[:])
+		if err != nil {
+			traceutil.AnnotateError(span, err)
+			return pubsub.ValidationIgnore
+		}
+
+		verified := blsSig.Verify(pKey, sigRoot[:])
+		if !verified {
+			return pubsub.ValidationReject
+		}
+		return pubsub.ValidationAccept
+	}
+}
+
+func rejectNilTopic(msg *pubsub.Message) validationFn {
+	return func(ctx context.Context) pubsub.ValidationResult {
+		if msg.Topic == nil {
+			return pubsub.ValidationReject
+		}
+		return pubsub.ValidationAccept
+	}
+}
+
+func ignoreEmptyCommittee(indices []types.CommitteeIndex) validationFn {
+	return func(ctx context.Context) pubsub.ValidationResult {
+		if len(indices) == 0 {
+			return pubsub.ValidationIgnore
+		}
+		return pubsub.ValidationAccept
+	}
+}
+
+func withValidationPipeline(ctx context.Context, fns ...validationFn) pubsub.ValidationResult {
+	for _, fn := range fns {
+		if result := fn(ctx); result != pubsub.ValidationAccept {
+			return result
+		}
+	}
```

Can this be done in parallel? 
Seems like an opportunity to split into multiple routines and fail fast if any methods fail.

---

No because it is critical pipelines are sequential in p2p validation, unfortunately